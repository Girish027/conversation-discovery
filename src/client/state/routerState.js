import { createSelector } from 'reselect';

// State, actions and reducer for the Router is managed by Connected Router.
// We are just adding selectors here.
export const stateKey = {
  location: 'location',
  hash: 'hash',
  pathname: 'pathname',
  search: 'search',
};

// SELECTORS
export const getRouterState = (state) => state.get('router');

export const getLocationSelector = createSelector(
  getRouterState,
  (routerState) => routerState.get(stateKey.location),
);

export const getSearchSelector = createSelector(
  getLocationSelector,
  (location) => location.get(stateKey.search),
);

export const getPathnameSelector = createSelector(
  getLocationSelector,
  (location) => location.get(stateKey.pathname),
);
