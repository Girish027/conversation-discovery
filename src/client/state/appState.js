import {
  fromJS, List, Map
} from 'immutable';
import { createSelector } from 'reselect';
import axios from 'axios';
import _ from 'lodash';
import { push } from 'connected-react-router';
import { getURLParams } from 'utils/URIUtils';
import { routeNames, getRoute } from 'utils/RouteHelper';
import { getPathnameSelector } from 'state/routerState';
import appLogger from 'utils/appLogger';
import { LOCATION_CHANGE, UPDATE_CAA } from 'state/types';
import { getAllProjects } from 'state/projectsState';
import { setDefaultsForAPI } from 'utils/api';
import { isUserAuthorized, getSortedClientsList } from 'utils/caaHelper';
import constants from '../constants';
import { clearStorage, loadStorage, saveStorage } from '../utils/storageManager';
import { logAmplitudeEvent } from '../utils/amplitudeUtils';
import { roles, updateRoleConfig } from '../utils/roleHelper';

export const types = {
  SET_USER_CONFIG: 'app/SET_USER_CONFIG',
  SET_CLIENTS_CONFIG: 'app/SET_CLIENTS_CONFIG',
  MODAL_IS_OPEN: 'app/MODAL_IS_OPEN',
  MODAL_MESSAGE: 'app/MODAL_MESSAGE',
  INIT_ANALYTICS: 'app/INIT_ANALYTICS',
  SET_UNIFIED_PORTAL_CLIENT_PICKER: 'app/SET_UNIFIED_PORTAL_CLIENT_PICKER',
  SHOW_WORD_CLOUD: 'app/SHOW_WORD_CLOUD',
};

export const stateKey = {
  authEnabled: 'authEnabled',
  clientId: 'clientId',
  componentClientId: 'componentClientId',
  appId: 'appId',
  accountId: 'accountId',
  userName: 'userName',
  userType: 'userType',
  itsBaseUrl: 'itsBaseUrl',
  analyticsKey: 'analyticsKey',
  clientConfigs: 'clientConfigs',
  modalIsOpen: 'modalIsOpen',
  modalMessage: 'modalMessage',
  modalState: 'modalState',
  unifiedPortalClientPicker: 'unifiedPortalClientPicker',
  docPortalUrl: 'docPortalUrl',
  unifiedPortalUrl: 'unifiedPortalUrl',
  contactSupportUrl: 'contactSupportUrl',
  oktaUserAccountUrl: 'oktaUserAccountUrl',
  isAnswers: 'isAnswers',
  isConversations: 'isConversations',
  userRole: 'userRole',
  roleConfig: 'roleConfig',
  showWordCloud: 'showWordCloud',
  wordCloudcontent: 'wordCloudcontent',
};

export const initialAppState = fromJS({
  authEnabled: false,
  clientId: '',
  componentClientId: '',
  appId: '',
  accountId: '',
  userName: '',
  userType: '',
  itsBaseUrl: '',
  analyticsKey: '',
  clientConfigs: List(),
  modalIsOpen: false,
  modalMessage: '',
  modalState: {},
  unifiedPortalClientPicker: false,
  docPortalUrl: '',
  unifiedPortalUrl: '',
  contactSupportUrl: '',
  oktaUserAccountUrl: '',
  isConversations: false,
  isAnswers: false,
  userRole: '',
  roleConfig: {},
  showWordCloud: false,
  wordCloudcontent: '',
});

const devlog = appLogger({ tag: 'appState' });

export const updateCAA = (caa) => ({
  type: UPDATE_CAA,
  caa,
});

export const setUnifiedPortalClientPicker = (unifiedPortalClientPicker) => ({
  type: types.SET_UNIFIED_PORTAL_CLIENT_PICKER,
  unifiedPortalClientPicker
});

export const setUserConfig = (config) => ({
  type: types.SET_USER_CONFIG,
  config,
});

export const setClientsConfig = (clientConfigs = []) => ({
  type: types.SET_CLIENTS_CONFIG,
  clientConfigs,
});

export const setShowWordCloud = (showWordCloud = false, wordCloudcontent = '') => ({
  type: types.SHOW_WORD_CLOUD,
  showWordCloud,
  wordCloudcontent,
});

export const setModalIsOpen = (modalIsOpen, modalState = {}) => ({
  type: types.MODAL_IS_OPEN,
  modalIsOpen,
  modalState,
});

export const setModalMessageState = (modalMessage) => ({
  type: types.MODAL_MESSAGE,
  modalMessage
});

export const setModalMessage = (modalMessage) => (dispatch) => {
  dispatch(setModalMessageState(modalMessage));
};

export const getConversationsConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.conversationsConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      const conversationsClientConfigs = configBody.clientConfigs;
      let clientConfigs = getClientConfigsSelector(getState());
      clientConfigs = clientConfigs.map((element) => {
        let configObj = element;
        const { clientId } = element;
        const conversationsClient = conversationsClientConfigs.find((config) => config.clientId === clientId);
        if (conversationsClient !== undefined) {
          configObj = { ...configObj, isConversations: true };
        }
        return configObj;
      });
      dispatch(setClientsConfig(clientConfigs));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

export const getAnswersConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.answersConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      const answersClientConfigs = configBody.clientConfigs;
      let clientConfigs = getClientConfigsSelector(getState());
      clientConfigs = clientConfigs.map((element) => {
        let configObj = element;
        const { clientId } = element;
        const answersClient = answersClientConfigs.find((config) => config.clientId === clientId);
        if (answersClient !== undefined) {
          configObj = { ...configObj, isAnswers: true };
        }
        return configObj;
      });
      dispatch(setClientsConfig(clientConfigs));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

export const getUserConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.userConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      setDefaultsForAPI();

      const sortClientConfigs = getSortedClientsList([...configBody.clientConfigs]);
      dispatch(setClientsConfig(sortClientConfigs));
      dispatch(getAnswersConfig());
      dispatch(getConversationsConfig());
      const userSelectedCAA = getCAASelector(getState());

      const { authorized, clientPicker, caa } = isUserAuthorized(userSelectedCAA, sortClientConfigs);
      if (clientPicker) {
        dispatch(setUnifiedPortalClientPicker(true));
      } else if (!authorized && (userSelectedCAA.clientId !== 'undefined' || userSelectedCAA.appId !== 'undefined')) {
        dispatch(setModalIsOpen(true, {
          modalName: constants.modals.unauthorized,
          header: 'Access Denied',
          message: `You are not authorized to view the client: ${userSelectedCAA.clientId}, app: ${userSelectedCAA.appId} and account: ${userSelectedCAA.accountId}`,
          onClickOk: () => dispatch(closeUnauthorizedModal()),
          onClickClose: () => dispatch(closeUnauthorizedModal()),
          onClickCancel: () => dispatch(closeUnauthorizedModal()),
        }));
      } else {
        dispatch(changeClient(caa));
      }
      dispatch(setUserConfig(configBody));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

const closeUnauthorizedModal = () => (dispatch) => {
  dispatch(setUnifiedPortalClientPicker(true)); dispatch(setModalIsOpen(false));
};

// When user changes client
export const changeClient = (caa) => (dispatch, getState) => {
  const state = getState();
  const existingCAA = getCAASelector(state);
  if (!_.isEqual(existingCAA, caa)) {
    dispatch(updateCAA(caa));
    clearStorage(constants.activeProjectId);
    clearStorage(constants.isConversations);
    clearStorage(constants.isAnswers);
    clearStorage(constants.userRole);
  }
  const { clientId, appId } = caa;
  const obj = {
    clientId, appId, toolId: constants.toolName, env: constants.environment
  };
  logAmplitudeEvent('IDTClients', obj);
  dispatch(loadRoute(caa));
  dispatch(getAllProjects(caa));
};

const loadRoute = (caa) => (dispatch, getState) => {
  // refresh the page only if it is HOMEPAGE/baseRoute.
  const state = getState();
  const currentPath = getPathnameSelector(state);
  if (currentPath === `${routeNames.HOME_PAGE}/` || currentPath === `${routeNames.BASE_ROUTE}/` || currentPath === `${routeNames.DISCOVER_INTENTS}`) {
    const projectId = loadStorage(constants.activeProjectId, '');
    const newRoute = getRoute(routeNames.DISCOVER_INTENTS, { ...caa, projectId });
    dispatch(push(newRoute));
  }
};

// REDUCERS
export const appReducer = (state = initialAppState, action) => {
  switch (action.type) {
    case types.SET_CLIENTS_CONFIG:
      return state.set(stateKey.clientConfigs, List(action.clientConfigs));
    case types.SET_USER_CONFIG:
      return state.merge(Map({
        userName: action.config.username,
        userType: action.config.userType,
        authEnabled: action.config.authEnabled,
        itsBaseUrl: action.config.itsBaseUrl,
        analyticsKey: action.config.analyticsKey,
        docPortalUrl: action.config.docPortalUrl,
        unifiedPortalUrl: action.config.unifiedPortalUrl,
        contactSupportUrl: action.config.contactSupportUrl,
        oktaUserAccountUrl: action.config.oktaUserAccountUrl,
      }));
    case types.SET_UNIFIED_PORTAL_CLIENT_PICKER:
      return state.set(stateKey.unifiedPortalClientPicker, action.unifiedPortalClientPicker);
    case types.MODAL_IS_OPEN:
      return state.merge(Map({
        modalIsOpen: action.modalIsOpen,
        modalState: action.modalState,
      }));
    case types.MODAL_MESSAGE:
      return state.set(stateKey.modalMessage, action.modalMessage);
    case types.SHOW_WORD_CLOUD:
      return state.merge(Map({
        showWordCloud: action.showWordCloud,
        wordCloudcontent: action.wordCloudcontent,
      }));
    // The below reducers are handling part of the global state change
    case UPDATE_CAA:
      return state.merge(Map(action.caa));
    case LOCATION_CHANGE: {
      const { search } = action.payload.location;
      if (search) {
        const {
          clientid, appid, accountid, componentClientId
        } = getURLParams(search);

        let isAnswers = false;
        let isConversations = false;

        const client = state.get(stateKey.clientId);
        const clientConfigs = state.get(stateKey.clientConfigs);
        const selectedClientConfig = clientConfigs.find((config) => config.clientId === clientid);
        let userRole = roles.unauthorized;
        if (selectedClientConfig) {
          userRole = selectedClientConfig.role;
        }

        if (selectedClientConfig && selectedClientConfig.isAnswers) {
          isAnswers = true;
        }
        if (selectedClientConfig && selectedClientConfig.isConversations) {
          isConversations = true;
        }

        if (client === '') {
          isConversations = loadStorage(constants.isConversations, false);
          isAnswers = loadStorage(constants.isAnswers, false);
          userRole = loadStorage(constants.userRole, roles.unauthorized);
        } else {
          saveStorage(constants.isConversations, isConversations);
          saveStorage(constants.isAnswers, isAnswers);
          saveStorage(constants.userRole, userRole);
        }
        const roleConfig = updateRoleConfig(userRole);
        return state.merge(Map({
          clientId: clientid,
          appId: appid,
          accountId: accountid,
          componentClientId,
          isAnswers,
          isConversations,
          userRole,
          roleConfig,
        }));
      }
      return state;
    }
    default:
      return state;
  }
};

// SELECTORS
export const getAppState = (state) => state.get('app');

export const getCAASelector = createSelector(
  getAppState,
  (appState) => ({
    clientId: appState.get(stateKey.clientId),
    appId: appState.get(stateKey.appId),
    accountId: appState.get(stateKey.accountId),
    componentClientId: appState.get(stateKey.componentClientId),
  })
);

export const getClientIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.clientId
);

export const getComponentClientIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.componentClientId
);

export const getAppIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.appId
);

export const getIsAnswersSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.isAnswers)
);

export const getIsConversationsSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.isConversations)
);

export const getAccountIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.accountId
);

export const getClientConfigsSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.clientConfigs)
);

export const getITSBaseUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.itsBaseUrl)
);

export const getShowWordCloudSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.showWordCloud)
);

export const getWordCloudContentSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.wordCloudcontent)
);

export const getModalOpenSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalIsOpen)
);

export const getModalStateSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalState)
);

export const getModalMessageSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalMessage)
);

export const getUnifiedPortalClientPicker = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.unifiedPortalClientPicker)
);

export const getDocPortalUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.docPortalUrl)
);

export const getUnifiedPortalUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.unifiedPortalUrl)
);

export const getUserTypeSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.userType)
);

export const getContactSupportUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.contactSupportUrl)
);

export const getUserRoleSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.userRole)
);

export const getRoleConfigSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.roleConfig)
);

export const getOktaUserAccountUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.oktaUserAccountUrl)
);
