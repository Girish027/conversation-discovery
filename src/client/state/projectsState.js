import {
  fromJS, List, Map
} from 'immutable';
import { createSelector } from 'reselect';
import _ from 'lodash';
import { getURLParams } from 'utils/URIUtils';
import { routeNames, getRoute } from 'utils/RouteHelper';
import { push } from 'connected-react-router';
import api from '../utils/api';
import Constants from '../constants';
import Language from '../Language';
import { getCAASelector, setModalIsOpen } from './appState';
import { headerActions } from './headerState';
import { getAllRuns, getRunsList, runActions } from 'state/runsState';
import { saveStorage } from '../utils/storageManager';
import { logAmplitudeEvent } from '../utils/amplitudeUtils';

import {
  LOCATION_CHANGE,
  UPDATE_CAA,
  SELECT_ACTIVE_PROJECT,
  SET_PROJECT_COUNT
} from './types';

export const types = {
  CLEAR_ALL_PROJECTS: 'CLEAR_ALL_PROJECTS',

  RECEIVE_ALL_PROJECTS: 'projects/RECEIVE_ALL_PROJECTS',
  RECEIVE_PROJECT: 'projects/RECEIVE_PROJECT',

  CREATE_PROJECT_START: 'projects/CREATE_PROJECT_START',
  CREATE_PROJECT_FAIL: 'projects/CREATE_PROJECT_FAIL',
  CREATE_PROJECT_SUCCESS: 'projects/CREATE_PROJECT_SUCCESS',

  EDIT_PROJECT_START: 'projects/EDIT_PROJECT_START',
  EDIT_PROJECT_FAIL: 'projects/EDIT_PROJECT_FAIL',
  EDIT_PROJECT_SUCCESS: 'projects/EDIT_PROJECT_SUCCESS',

  DELETE_PROJECT_START: 'projects/DELETE_PROJECT_START',
  DELETE_PROJECT_FAIL: 'projects/DELETE_PROJECT_FAIL',
  DELETE_PROJECT_SUCCESS: 'projects/DELETE_PROJECT_SUCCESS',

};

export const stateKey = {
  projects: 'projects',
  activeProjectId: 'activeProjectId',
  trackProjects: 'trackProjects',
  projectRuns: 'projectRuns',
  projectCreationInProgress: 'projectCreationInProgress',
  projectCount: 'projectCount',
};

const projectIdStr = 'projectId';

export const initialProjectState = fromJS({
  // should contain list of projects with all its metadata including dataset name and id
  // this is a Immutable List that contains Maps
  projects: List(),
  // id of project currently selected by the user
  activeProjectId: '',
  // indicates whether any project is being validated and hence to be tracked.
  trackProjects: false,
  // indicates whether a project is being created. Used in create project modal
  projectCreationInProgress: false,
  // indicated whether a project is being loaded.
  projectCount: -1,
});

// ACTIONS
export const createProjectStart = (project) => ({
  type: types.CREATE_PROJECT_START,
  project,
});

export const createProjectFail = (error) => ({
  type: types.CREATE_PROJECT_FAIL,
  error,
});

export const createProjectSuccess = (project) => ({
  type: types.CREATE_PROJECT_SUCCESS,
  project,
});

export const onCreateProjectFail = (error = {}) => (dispatch) => {
  const { err, message } = error;

  dispatch(createProjectFail(Language[err]));
  const projectFailure = {
    toolId: Constants.toolName,
    projectStatus: Constants.notification.error,
    env: Constants.environment,
  };
  logAmplitudeEvent('IDT_ProjectFailureEvent', projectFailure);
  dispatch(setModalIsOpen(true, {
    modalName: Constants.modals.createProject,
    error: message,
    errorType: Language[err] ? Language[err] : message,
  }));
};

export const onCreateProjectSuccess = (response = {}) => (dispatch, getState) => {
  const caa = getCAASelector(getState());
  dispatch(createProjectSuccess(response));
  dispatch(selectProject(response.projectId));
  dispatch(setModalIsOpen(false));
  dispatch(
    headerActions.showNotificationWithTimer(
      Language.NOTIFICATIONS.projectCreated(response.projectName),
      Constants.notification.success,
    )
  );
  dispatch(projectActions.getAllProjects(caa));
  // TODO : log amplitude events
  const projectSuccess = {
    toolId: Constants.toolName,
    projectId: response.projectId,
    projectName: response.projectName,
    projectStatus: Constants.notification.success,
    env: Constants.environment,
  };
  logAmplitudeEvent('IDT_ProjectSuccessEvent', projectSuccess);
};

export const createProject = (data) => (dispatch, getState) => {
  const url = Constants.serverApiUrls.projects(data);

  const {
    projectName, datasetFile, datasetName, projectDescription,
  } = data;
  dispatch(createProjectStart({ projectName }));
  const createProjectEvent = {
    toolId: Constants.toolName,
    projectName,
    datasetName,
    projectDescription,
    env: Constants.environment,
  };
  logAmplitudeEvent('IDT_CreateProjectEvent', createProjectEvent);
  // TODO: get param names from swagger
  const formData = new FormData();
  formData.append('projectName', projectName);
  formData.append('projectDescription', projectDescription);
  formData.append('datasetName', datasetName);
  formData.append('upfile', datasetFile, datasetName);

  api.post({
    dispatch,
    getState,
    url,
    data: formData,
    onApiSuccess: onCreateProjectSuccess,
    onApiError: onCreateProjectFail,
  });
};

export const receiveProject = (project) => ({
  type: types.RECEIVE_PROJECT,
  project
});

export const onGetProjectFail = (error = {}) => (dispatch) => {
  const { err, message } = error;

  const errorMsg = Language[err] ? Language[err] : message;
  dispatch(headerActions.showNotificationWithTimer(errorMsg, Constants.notification.error));
};

export const getProject = (data) => (dispatch, getState) => {
  const url = Constants.serverApiUrls.oneProject(data);
  api.get({
    dispatch,
    getState,
    url,
    onApiSuccess: receiveProject,
    onApiError: onGetProjectFail,
  });
};

export const receiveAllProjects = (projectsList = []) => ({
  type: types.RECEIVE_ALL_PROJECTS,
  projectsList
});

export const onGetAllProjectsSuccess = (projectsList = []) => (dispatch, getState) => {
  dispatch(receiveAllProjects(projectsList));
  const state = getState();
  const activeProjectId = getActiveProjectId(state);
  const runsList = getRunsList(state);
  // load runs only if the state does not have them.
  if (activeProjectId && runsList.size === 0) {
    dispatch(getAllRuns());
  }
};

export const getAllProjects = (data) => (dispatch, getState) => {
  const url = Constants.serverApiUrls.projects(data);
  api.get({
    dispatch,
    getState,
    url,
    onApiSuccess: onGetAllProjectsSuccess,
    onApiError: onGetProjectFail,
  });
};

export const setActiveProject = (activeProjectId) => ({
  type: SELECT_ACTIVE_PROJECT,
  activeProjectId
});

export const setProjectCount = (projectCount) => ({
  type: SET_PROJECT_COUNT,
  projectCount
});

export const selectProject = (projectId) => (dispatch, getState) => {
  dispatch(setActiveProject(projectId));
  dispatch(getAllRuns());
  const caa = getCAASelector(getState());
  const newRoute = getRoute(routeNames.DISCOVER_INTENTS, { ...caa, projectId });
  dispatch(push(newRoute));

  saveStorage(Constants.activeProjectId, projectId);
};

export const editProjectStart = (project) => ({
  type: types.EDIT_PROJECT_START,
  project,
});

export const editProjectFail = (error) => ({
  type: types.EDIT_PROJECT_FAIL,
  error,
});

export const editProjectSuccess = (project) => ({
  type: types.EDIT_PROJECT_SUCCESS,
  project,
});

export const onEditProjectFail = (projectId, data, systemProject, error = {}) => (dispatch) => {
  const { err, message } = error;
  dispatch(editProjectFail(Language[err]));
  if (!systemProject) {
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.editProject,
      projectId,
      formData: data,
      error: message,
      errorType: Language[err] ? Language[err] : message,
    }));
  } else if (err === Constants.PROJECT_ALREADY_EXISTS) {
    dispatch(headerActions.showNotificationWithTimer(Language.SYSTEM_PROJECT_RENAME, Constants.notification.error));
  } else {
    dispatch(headerActions.showNotificationWithTimer(message, Constants.notification.error));
  }
};

export const onEditProjectSuccess = (runId, systemAccess, response = {}) => (dispatch) => {
  dispatch(setModalIsOpen(false));
  dispatch(editProjectSuccess(response));
  const { projectId, projectName } = response;
  const editProject = {
    toolId: Constants.toolName, projectId, projectName, env: Constants.environment
  };
  logAmplitudeEvent('IDT_EditProject', editProject);
  if (systemAccess) {
    dispatch(
      headerActions.showNotificationWithTimer(
        Language.NOTIFICATIONS.projectSysEdited(Constants.LIVE_DATA_ANALYSIS),
        Constants.notification.success
      ));
  } else {
    dispatch(
      headerActions.showNotificationWithTimer(
        Language.NOTIFICATIONS.projectEdited(response.projectName),
        Constants.notification.success
      ));
  }
  if (runId) {
    dispatch(runActions.selectRun(runId));
  }
};

export const editProject = (projectId, data, systemAccess, runId) => (dispatch, getState) => {
  const state = getState();
  const caa = getCAASelector(state);
  let headers;
  const url = Constants.serverApiUrls.oneProject({ ...caa, projectId });
  if (systemAccess) {
    headers = { systemAccess };
  } else {
    headers = { systemAccess: false };
  }

  dispatch(editProjectStart({ data, projectId, caa }));

  api.patch({
    dispatch,
    getState,
    url,
    headers,
    data,
    onApiSuccess: (response) => onEditProjectSuccess(runId, systemAccess, response),
    onApiError: (error) => onEditProjectFail(projectId, data, systemAccess, error),
  });
};

export const deleteProjectStart = (project) => ({
  type: types.DELETE_PROJECT_START,
  project,
});

export const deleteProjectFail = (error) => ({
  type: types.DELETE_PROJECT_FAIL,
  error,
});

export const deleteProjectSuccess = (project) => ({
  type: types.DELETE_PROJECT_SUCCESS,
  project,
});

export const onDeleteProjectSuccess = (response = {}) => (dispatch, getState) => {
  const { projectName, projectId } = response;
  dispatch(deleteProjectSuccess(response));
  dispatch(selectProject(getActiveProjectId(getState())));
  const deleteProject = {
    toolId: Constants.toolName,
    projectId,
    projectName,
    env: Constants.environment,
  };
  logAmplitudeEvent('IDT_DeleteProject', deleteProject);
  // dispatch(getAllRuns())
  dispatch(headerActions.showNotificationWithTimer(
    Language.NOTIFICATIONS.projectDeleted(projectName),
    Constants.notification.success,
  ));
};

export const deleteProject = (projectId) => (dispatch, getState) => {
  const state = getState();
  const caa = getCAASelector(state);
  const url = Constants.serverApiUrls.oneProject({ ...caa, projectId });
  dispatch(deleteProjectStart({ caa, projectId }));

  api.delete({
    dispatch,
    getState,
    url,
    onApiSuccess: onDeleteProjectSuccess,
    onApiError: deleteProjectFail,
  });
};

// REDUCERS
export const projectReducer = (state = initialProjectState, action) => {
  switch (action.type) {
    case types.CLEAR_ALL_PROJECTS: return initialProjectState;

    case types.CREATE_PROJECT_START:
      return state.set(stateKey.projectCreationInProgress, true);

    case types.CREATE_PROJECT_FAIL:
      return state.set(stateKey.projectCreationInProgress, false);

    case types.RECEIVE_ALL_PROJECTS: {
      const { projectsList = [] } = action;
      let trackProjects = false;
      const projectInProgress = projectsList.find((project) => project.status !== Constants.status.ENABLED);
      if (!_.isNil(projectInProgress)) {
        trackProjects = true;
      }

      // select latest project if none or invalid is selected
      let activeProjectId = state.get(stateKey.activeProjectId);
      if (activeProjectId) {
        const valid = projectsList.some((pro) => pro.projectId === activeProjectId);
        if (!valid) {
          activeProjectId = '';
        }
      }
      if (!activeProjectId && projectsList.length) {
        activeProjectId = projectsList[0].projectId;
      }

      return state
        .set(stateKey.projects, fromJS(projectsList))
        .set(stateKey.trackProjects, trackProjects)
        .set(stateKey.activeProjectId, activeProjectId)
        .set(stateKey.projectCount, projectsList.length);
    }
    case types.CREATE_PROJECT_SUCCESS:
    case types.EDIT_PROJECT_SUCCESS:
    case types.RECEIVE_PROJECT: {
      let trackProjects = state.get(stateKey.trackProjects);
      let activeProjectId = state.get(stateKey.activeProjectId);

      const { projectId } = action.project;
      const projects = state.get(stateKey.projects);
      const projectIndex = projects.findIndex((proj) => proj.get(projectIdStr) === projectId);
      let newProjects;
      if (projectIndex !== -1) {
        newProjects = projects.set(projectIndex, Map(action.project));
      } else {
        newProjects = projects.insert(0, Map(action.project));
        // select this project if this is the first project created
        if (projects.size === 1) {
          activeProjectId = action.project.projectId;
          trackProjects = true;
        }
      }
      return state.set(stateKey.projects, newProjects)
        .set(stateKey.trackProjects, trackProjects)
        .set(stateKey.activeProjectId, activeProjectId)
        .set(stateKey.projectCreationInProgress, false)
        .set(stateKey.projectCount, newProjects.size);
    }

    case types.DELETE_PROJECT_SUCCESS: {
      const deletedProject = action.project;
      const { projectId } = deletedProject;
      const projects = state.get(stateKey.projects);
      // Remove from the list of projects
      const indexOfProjectToBeRemoved = projects.findIndex((project) => project.get(projectIdStr) === projectId);
      const remainingProjects = (indexOfProjectToBeRemoved !== -1) ? projects.delete(indexOfProjectToBeRemoved) : projects;

      // Select the first project if this deleted project was active project
      let activeProjectId = state.get(stateKey.activeProjectId);
      if (activeProjectId === projectId) {
        if (remainingProjects.size) {
          activeProjectId = remainingProjects.get(0).get(projectIdStr);
        } else {
          activeProjectId = '';
        }
      }
      return state
        .set(stateKey.projects, remainingProjects)
        .set(stateKey.activeProjectId, activeProjectId)
        .set(stateKey.projectCount, remainingProjects.size);
    }

    // The below reducers are handling part of the global state change
    case SELECT_ACTIVE_PROJECT: {
      return state.set(stateKey.activeProjectId, action.activeProjectId);
    }

    case LOCATION_CHANGE: {
      const { search } = action.payload.location;
      const { project: projectId } = getURLParams(search);
      return state.set(stateKey.activeProjectId, projectId);
    }

    case UPDATE_CAA:
      // clear projects state when CAA changes
      return initialProjectState;

    default:
      return state;
  }
};

// SELECTORS
export const getProjectsState = (state) => state.get('projects');

export const getActiveProjectId = (state) => getProjectsState(state).getIn([stateKey.activeProjectId]);

export const getProjectCount = (state) => getProjectsState(state).getIn([stateKey.projectCount]);
export const getProjectsToTrackSelector = (state) => getProjectsState(state).getIn([stateKey.trackProjects]);
export const getProjectCreationInProgress = (state) => getProjectsState(state).getIn([stateKey.projectCreationInProgress]);

export const getProjectsList = createSelector(
  getProjectsState,
  (projectsState) => projectsState.get(stateKey.projects),
);

export const getActiveProject = createSelector(
  getActiveProjectId, getProjectsList,
  (activeProjectId, projectsList) => projectsList.find((p) => p.get(projectIdStr) === activeProjectId),
);

// SELECTORS TO BE USED IN COMPONENTS
export const projectSelectors = {
  getProjectsList,
  getActiveProject,
  getActiveProjectId,
  getProjectsToTrackSelector,
  getProjectCreationInProgress,
  getProjectCount,
};

// ACTIONS TO BE DISPATCHED FROM COMPONENTS
export const projectActions = {
  createProject,
  getProject,
  getAllProjects,
  selectProject,
  deleteProject,
  editProject,
};
