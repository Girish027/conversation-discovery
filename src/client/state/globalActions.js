import { downloadFile } from 'utils/FileUtils';
import { getCAASelector } from 'state/appState';
import { getActiveProjectId } from 'state/projectsState';
import Constants from '../constants';
import { logAmplitudeEvent } from '../utils/amplitudeUtils';
import {
  DOWNLOAD_FILE
} from './types';

export const downloadResults = (constructURL, runId, clusterId = '') => (dispatch, getState) => {
  const state = getState();
  const projectId = getActiveProjectId(state);
  const caa = getCAASelector(state);
  const locationUrl = constructURL({
    ...caa, projectId, runId, clusterId
  });
  dispatch({ type: DOWNLOAD_FILE, locationUrl });
  const downloadRun = { toolId: Constants.toolName, runId, env: Constants.environment };
  logAmplitudeEvent('IDT_DownloadRunResults', downloadRun);
  downloadFile(locationUrl);
};
