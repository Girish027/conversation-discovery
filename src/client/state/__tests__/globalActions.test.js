import constants from 'components/../constants';
import { downloadResults } from '../globalActions';
import * as FileUtils from 'utils/FileUtils';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import * as projectsState from 'state/projectsState';
import * as appState from 'state/appState';
import {
  fromJS, List
} from 'immutable';

let store;
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
store = mockStore(fromJS({
  auth: {}
      }));

jest.mock('utils/FileUtils');
jest.mock('state/projectsState');
jest.mock('state/appState');

describe('state/globalActions:', function () {
  beforeAll(() => {
    FileUtils.downloadFile = jest.fn();
    projectsState.getActiveProjectId = jest.fn();
    appState.getCAASelector = jest.fn();
  });

  describe('downloadResults' , function() {

    test('should download File', () => {
      FileUtils.downloadFile.mockReturnValue('downloadFile');
      projectsState.getActiveProjectId.mockReturnValue('getActiveProjectId');
      appState.getCAASelector.mockReturnValue('getCAASelector');
      let runId = 'run-1';
      store.dispatch(downloadResults(constants.serverApiUrls.downloadRun, runId));
      expect(store.getActions()).toMatchSnapshot();
      expect(FileUtils.downloadFile).toHaveBeenCalled();
    });

  });

});