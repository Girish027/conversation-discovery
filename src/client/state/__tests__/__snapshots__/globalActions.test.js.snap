// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state/globalActions: downloadResults should download File 1`] = `
Array [
  Object {
    "locationUrl": "//localhost:8080/conversationdiscovery/clients/undefined/accounts/undefined/applications/undefined/projects/getActiveProjectId/runs/run-1/results?type=original",
    "type": "all/DOWNLOAD_FILE",
  },
]
`;
